### ISSUE TITLE
[Caloris]  Github Code

### ISSUE THREAD

**Issue by spolu** [dust-tt/dust#2324]
- Add code (ability to select code or issues/discussions)
- What about .md outside of code?
- What about commits?

**Comment by spolu** [comment]
For reference:
- https://dust.tt/w/0ec9852c2f/assistant/8fd95b8f8f
- https://github.com/PyGithub/PyGithub/blob/main/github/Repository.py#L2148

**Comment by fontanierh** [comment]
I have not looked into ways to pull repo content so far, so I don't have any prior here. 
But reading your post, I tend to agree that the tarball looks like an easier (and likely less flaky) solution.

Repos probably aren't that big anyway, so downloading the whole thing doesn't sound too bad.

**Comment by fontanierh** [comment]
Is the plan to index the content in our existing semantic search directly or to design a new type of retrieval/index specific for code ?

**Comment by spolu** [comment]
Regular semantic search.

Advanced use-cases can use Dust apps to work on full files :+1:

**Comment by spolu** [comment]
repos can be big IMHO but we can also have limits in place of course

**Comment by fontanierh** [comment]
> repos can be big IMHO but we can also have limits in place of course

I think that we'll likely want to exclude any file that is more than some reasonable limit (1mb is already a lot for a code file no ?)
Is there a way to get a filtered tarball, excluding any big file ?

**Comment by spolu** [comment]
Not sure if there is but that seems fine we can abort at 1GB and then filter as we explore the repo I presume

**Comment by spolu** [comment]
# Proposal

- Update Git repo code once a day on the repository default branch
- Update consists in downloading and filtering tar ball + doing incremental update on changed files based on their hash
- The github connection will show {repo}/issue {repo}/discussions and {repo}/code/[[files]] with granular selection of directories when adding as a data source.

## Alternatives

The alternative would be to sync in realtime when we receive a webhook that some code was merged to main. This is actually something we can build on top if the need arise. But we can't solely base ourselves on that.

For the initial sync we could use an API but there seems to be limits for large repos + that's a whole lot of API interactions to traverse a full repo that is quite big.

**Comment by philipperolet** [comment]
TLDR :+1: 
The tar download solution SGTM, seems better / simpler / less flaky than alternatives as y'all say; and we can easily increase frequency if it seems warranted by usage (e.g move to every hour if repos really want so) => can somewhat scale for a while
FUD #1 -> code mixed with non-code in sem search -> not that risky though because semantic separation should work
(would repeat file title in structure, and maybe even function later down the way)
FUD #2 -> crap in repos e.g large non-code files etc -> well adressed by the 1GB cut but to monitor; Dust repo is 150MB and we mostly avoid crap so I fear other repos might not be as clean, however solutions seem within reach even if we realize we commonly have 3GB repos so ok fine

**Comment by spolu** [comment]
Agreed with FUD #2 we'll def start concervative. Strict cap on file size to be ingested + whitelist of file extensions :+1:

**Comment by spolu** [comment]
That being said I think we should start with something very lame. That's what Alan is using today and they seem already quite happy.

**Comment by spolu** [comment]
Tasks will consists in:

- Move issues/discussions to separate parents in github connectors (requires migration)
- Add capability to sync code behind feature flag
- Frame product rollout

**Comment by spolu** [comment]
Framing complete, issues created referring to this issue.
